x86_symbolic/src/Data/Macaw/X86/Symbolic/ABI/SysV.hs

```diff
@@ -0,0 +1,219 @@
+{-|
+Copyright        : (c) Galois, Inc 2024
+Maintainer       : Langston Barrett <langston@galois.com>
+
+On x86_64 with the SysV ABI, the stack grows \"downwards\" from high addresses
+to low. The end of the stack is initialized with the ELF auxiliary vector (not
+modelled here), and functions expect the following data to be available above
+their stack frame (i.e., just above the address in @rsp@), from high addresses
+to low:
+
+* Their stack-spilled arguments
+* The return address
+
+The functions in this module support manipulation of a stack under these
+constraints. ABI-compatible machine code translated by Macaw manages the stack
+for itself, so this module is primarily helpful for initial setup of the stack,
+before starting symbolic execution.
+
+Helpful links:
+
+* <https://eli.thegreenplace.net/2011/09/06/stack-frame-layout-on-x86-64>
+* <https://wiki.osdev.org/System_V_ABI#x86-64>
+
+TODO: The stack is supposed to be 16-byte aligned before a @call@.
+
+This module is meant to be imported qualified.
+-}
+
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE ImplicitParams #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+
+module Data.Macaw.X86.Symbolic.ABI.SysV
+  ( StackPointer
+  , getStackPointer
+  , stackPointerReg
+  , RetAddr(..)
+  , freshRetAddr
+  , SpilledArgs(..)
+  , writeSpilledArgs
+  , writeRetAddr
+  , allocStackFrame
+  , pushStackFrame
+  , allocStack
+  ) where
+
+import Control.Lens qualified as Lens
+import Control.Monad qualified as Monad
+import Data.BitVector.Sized qualified as BVS
+import Data.Macaw.Symbolic qualified as MS
+import Data.Macaw.Symbolic.Stack qualified as MSS
+import Data.Macaw.X86 qualified as X86
+import Data.Macaw.X86.Symbolic qualified as X86S
+import Data.Parameterized.Classes (ixF')
+import Data.Parameterized.Context qualified as Ctx
+import Data.Sequence qualified as Seq
+import Lang.Crucible.Backend qualified as C
+import Lang.Crucible.LLVM.Bytes qualified as Bytes
+import Lang.Crucible.LLVM.DataLayout qualified as CLD
+import Lang.Crucible.LLVM.MemModel qualified as MM
+import Lang.Crucible.Simulator qualified as C
+import What4.Interface qualified as WI
+
+-- | Helper, not exported
+ptrBytes :: Integer
+ptrBytes = 8
+
+ptrRepr :: WI.NatRepr 64
+ptrRepr = WI.knownNat
+
+-- | A pointer to a SysV-compatible x86_64 stack
+newtype StackPointer sym = StackPointer { getStackPointer :: MM.LLVMPtr sym 64 }
+
+-- | A lens for accessing the stack pointer register as a 'StackPointer'
+stackPointerReg ::
+  Lens.Lens'
+    (Ctx.Assignment (C.RegValue' sym) (MS.MacawCrucibleRegTypes X86.X86_64))
+    (StackPointer sym)
+stackPointerReg =
+  Lens.lens
+    (\regs -> StackPointer (C.unRV (regs Lens.^. ixF' X86S.rsp)))
+    (\regs v -> regs Lens.& ixF' X86S.rsp Lens..~ C.RV (getStackPointer v))
+
+-- | A return address
+newtype RetAddr sym = RetAddr { getRetAddr :: MM.LLVMPtr sym 64 }
```

Comment:
Is there ever a scenario where you'd want the block number for the return address's `LLVMPtr` to be non-zero? If not, wouldn't it be more direct to represent this as a `SymBV sym 64` instead?

Reply:
I can't think of one right now, but perhaps in the future we would support something like `mmap`ing new code allocations, e.g., in the context of JIT compilation, and would support `ret`ing to such allocations? I'm not sure there's a super clear advantage to disallowing it, other than making the more common usage (block number 0) more obvious.